### PR TITLE
acom cal

### PR DESCRIPTION
**Pull request by bsavitzky**
Refactors to streamline calibration - see new notebook in py4DSTEM_tutorials/notebooks/acom/ACOM_02_sim_cal.ipynb for demo code.

This update is fully functional, however some documentation remains incomplete.

### CODE REVIEW

File: py4DSTEM/io/datastructure/datacube.py [hunk 1]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.R_Nx = self.R_Nx
+        self.coordinates.R_Ny = self.R_Ny
 
     def swap_RQ(self):
         """
         Swap real and reciprocal space coordinates.
         """
         self = preprocess.swap_RQ(self)
         self.update_slice_parsers()
+        self.coordinates.Q_Nx = self.Q_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:737]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 2]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.R_Nx = self.R_Nx
+        self.coordinates.R_Ny = self.R_Ny
 
     def swap_Rxy(self):
         """
         Swap real space x and y coordinates.
         """
         self = preprocess.swap_Rxy(self)
         self.update_slice_parsers()
+        self.coordinates.Q_Nx = self.Q_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:748]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 3]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.Q_Ny = self.Q_Ny
+        self.coordinates.R_Nx = self.R_Nx
+        self.coordinates.R_Ny = self.R_Ny
 
     def swap_Qxy(self):
         """
         Swap reciprocal space x and y coordinates.
         """
         self = preprocess.swap_Qxy(self)
+        Q_Nx,Q_Ny = self.coordinates.Q_Nx,self.coordinates.Q_Ny
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:758]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 4]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.Q_Nx = self.Q_Nx
+        self.coordinates.Q_Ny = self.Q_Ny
 
     def crop_data_diffraction(self,crop_Qx_min,crop_Qx_max,crop_Qy_min,crop_Qy_max):
         self = preprocess.crop_data_diffraction(self,crop_Qx_min,crop_Qx_max,crop_Qy_min,crop_Qy_max)
+        self.coordinates.Q_Nx = self.Q_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:764]: See above.

File: py4DSTEM/io/datastructure/datacube.py [hunk 5]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.Q_Ny = self.Q_Ny
 
     def crop_data_real(self,crop_Rx_min,crop_Rx_max,crop_Ry_min,crop_Ry_max):
         self = preprocess.crop_data_real(self,crop_Rx_min,crop_Rx_max,crop_Ry_min,crop_Ry_max)
+        self.coordinates.R_Nx = self.R_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:769]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 6]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.R_Ny = self.R_Ny
 
     def bin_data_diffraction(self, bin_factor):
         self = preprocess.bin_data_diffraction(self, bin_factor)
+        self.coordinates.Q_Nx = self.Q_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:774]: See abovev

File: py4DSTEM/io/datastructure/datacube.py [hunk 7]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.Q_Ny = self.Q_Ny
 
     def bin_data_mmap(self, bin_factor, dtype=np.float32):
         self = preprocess.bin_data_mmap(self, bin_factor, dtype=dtype)
+        self.coordinates.Q_Nx = self.Q_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:779]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 8]
```diff
@@ -65,41 +725,98 @@ def set_scan_shape(self,R_Nx,R_Ny):
...
+        self.coordinates.Q_Ny = self.Q_Ny
 
     def bin_data_real(self, bin_factor):
         self = preprocess.bin_data_real(self, bin_factor)
+        self.coordinates.R_Nx = self.R_Nx
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:784]: See above

File: py4DSTEM/io/datastructure/datacube.py [hunk 9]
```diff
@@ -129,22 +846,22 @@ def get_virtual_image_rect_integrate(self,slice_x,slice_y):
         Returns a virtual image as an ndarray, generated from a rectangular detector in integration
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_rect_integrate(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_rect_integrate(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:849]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 10]
```diff
@@ -129,22 +846,22 @@ def get_virtual_image_rect_integrate(self,slice_x,slice_y):
...
 
     def get_virtual_image_circ_integrate(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in integration
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_circ_integrate(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_circ_integrate(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:856]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 11]
```diff
@@ -129,22 +846,22 @@ def get_virtual_image_rect_integrate(self,slice_x,slice_y):
...
 
     def get_virtual_image_annular_integrate(self,slice_x,slice_y,R):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in integration
         mode. Also returns a bool indicating success or failure. The input parameter R is the ratio
         of the inner to the outer detector radii.
         """
-        return virtualimage.get_virtual_image_annular_integrate(self,slice_x,slice_y,R)
+        return virutalimage_viewer.get_virtual_image_annular_integrate(self,slice_x,slice_y,R)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:864]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 12]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
         Returns a virtual image as an ndarray, generated from a rectangular detector in difference
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_rect_diffX(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_rect_diffX(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:873]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 13]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
...
 
     def get_virtual_image_rect_diffY(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a rectangular detector in difference
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_rect_diffY(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_rect_diffY(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:880]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 14]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
...
 
     def get_virtual_image_circ_diffX(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in difference
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_circ_diffX(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_circ_diffX(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:887]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 15]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
...
 
     def get_virtual_image_circ_diffY(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in difference
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_circ_diffY(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_circ_diffY(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:894]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 16]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
...
 
     def get_virtual_image_annular_diffX(self,slice_x,slice_y,R):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in difference
         mode. Also returns a bool indicating success or failure. The input parameter R is the ratio
         of the inner to the outer detector radii.
         """
-        return virtualimage.get_virtual_image_annular_diffX(self,slice_x,slice_y,R)
+        return virutalimage_viewer.get_virtual_image_annular_diffX(self,slice_x,slice_y,R)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:902]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 17]
```diff
@@ -153,44 +870,44 @@ def get_virtual_image_rect_diffX(self,slice_x,slice_y):
...
 
     def get_virtual_image_annular_diffY(self,slice_x,slice_y,R):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in difference
         mode. Also returns a bool indicating success or failure. The input parameter R is the ratio
         of the inner to the outer detector radii.
         """
-        return virtualimage.get_virtual_image_annular_diffY(self,slice_x,slice_y,R)
+        return virutalimage_viewer.get_virtual_image_annular_diffY(self,slice_x,slice_y,R)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:910]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 18]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
         Returns a virtual image as an ndarray, generated from a rectangular detector in CoM
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_rect_CoMX(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_rect_CoMX(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:919]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 19]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
...
 
     def get_virtual_image_rect_CoMY(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a rectangular detector in CoM
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_rect_CoMY(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_rect_CoMY(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:926]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 20]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
...
 
     def get_virtual_image_circ_CoMX(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in CoM
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_circ_CoMX(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_circ_CoMX(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:933]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 21]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
...
 
     def get_virtual_image_circ_CoMY(self,slice_x,slice_y):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in CoM
         mode. Also returns a bool indicating success or failure.
         """
-        return virtualimage.get_virtual_image_circ_CoMY(self,slice_x,slice_y)
+        return virutalimage_viewer.get_virtual_image_circ_CoMY(self,slice_x,slice_y)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:940]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 22]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
...
 
     def get_virtual_image_annular_CoMX(self,slice_x,slice_y,R):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in CoM
         mode. Also returns a bool indicating success or failure. The input parameter R is the ratio
         of the inner to the outer detector radii.
         """
-        return virtualimage.get_virtual_image_annular_CoMX(self,slice_x,slice_y,R)
+        return virutalimage_viewer.get_virtual_image_annular_CoMX(self,slice_x,slice_y,R)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:948]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/io/datastructure/datacube.py [hunk 23]
```diff
@@ -199,44 +916,44 @@ def get_virtual_image_rect_CoMX(self,slice_x,slice_y):
...
 
     def get_virtual_image_annular_CoMY(self,slice_x,slice_y,R):
         """
         Returns a virtual image as an ndarray, generated from a circular detector in CoM
         mode. Also returns a bool indicating success or failure. The input parameter R is the ratio
         of the inner to the outer detector radii.
         """
-        return virtualimage.get_virtual_image_annular_CoMY(self,slice_x,slice_y,R)
+        return virutalimage_viewer.get_virtual_image_annular_CoMY(self,slice_x,slice_y,R)
```
Comment (sezelt) [inline py4DSTEM/io/datastructure/datacube.py:956]: `NameError: name 'virutalimage_viewer' is not defined`

File: py4DSTEM/process/virtualimage/virtualimage.py [hunk 24]
```diff
@@ -85,7 +78,36 @@ def get_virtualimage_ann(datacube, x0, y0, Ri, Ro, verbose=True):
...
 
+    Args:
+        datacube (DataCube):    Input datacube with dimensions (R_Nx, R_Nx, Q_Nx, Q_Ny)
+        mask (bool):            Mask with dimensions (Q_Nx, Q_Ny)
+        verbose (bool):         Use progress bar
 
+    Returns:
+        (2D array): the virtual image
+    """
+    assert isinstance(datacube, DataCube)
```
Comment (sezelt) [inline py4DSTEM/process/virtualimage/virtualimage.py:97]: `NameError: name 'DataCube' is not defined`

**Review by sezelt — COMMENTED** [review]
Overall, I like the approach of attaching the `Coordinates` object and some of the calibration to the `DataCube` object itself. However, I think that this PR takes this a little too far, attaching too much of the analysis to the `DataCube` object. Much of the processing that this is attaching to the `DataCube` really belongs elsewhere, in our other `DataObject`s. For example, BVM related functionality should be handled in `PointListArray`s and `DiffarctionSlice`s. I think even adding some new objects would clean things up massively: in particular creating a `Detector` object that can tie together all the virtual detector creation and virtual image generation, and a `BraggVectorMap` object that ties together all the calibration performed using the BVM. Ideally, objects like the virtual images, `BraggVectorMap`, etc., will share the same `Coordinates` object as the `DataCube` that yielded them, so that calibrations can be tied together as well. These are major changes, though, so at this point it seems like we need to discuss this together and come to some consensus. 

The implementation of the `DataCube`-centric analysis pipeline is also quite buggy in the current form. One particular frustration is that many of the calibration functions set a flag after they've been run, which prevents them from being run again... so you can't tweak any of the calibrations once they've been run once! The order of operations is also too rigidly fixed, so for example if virtual images haven't been generated you can't run disk detection. 

I am also concerned that there does not seem to be a simple way to save the analysis performed using this pipeline. The save function has not been touched, so I assume that all of the virtual images, probe kernel, Bragg disks, etc., that are attached to the `DataCube` are not saved when you use `py4DSTEM.io.save`. This also needs to be fixed before we can merge this. 

This PR also seems to contain a substantial refactor of the ellipse routines, and a major reorganization of the ptychography submodule. I haven't reviewed them yet, as they should really each be split into their own PR, but just from skimming through the changes they seem to be in good shape.

**Comment by bsavitzky** [comment]
Thanks for all the comments, Steve!  Very helpful, and I agree with most of them.  The rigid flag structure involved in calibrating the Bragg peak positions is particularly silly, and is this way now because I think the right way to do it is to refactor the PointListArray object, which would be a more major overhauls, or maybe make a new derivative class.  Anyway agreed that this has to change, this is definitely a temporary approach.

I do agree that not everything should live inside a DataCube, however, I think there is good reason to nest derivative data inside of them.  So, moving some functionality into PointListArrays / DiffractionSlice / RealSlice might make sense, but frequently I think does not - e.g. in measure_origin, while the output could be some RealSlice, we are taking a measurement of the DataCube, so it should be a DataCube, and the output should be attached to the DataCube.  Similarly, disk positions computed from a DataCube should be attached to the DataCube.  Methods that process those disk positions further - e.g. computing a BVM - could then live inside the disks PointListArray, or in the DataCube....this is a little trickier and we can discuss more in person.

The elliptical calibration changes are a separate PR - #202, submitted Sept 30, and not yet merged.  I build on top of that because it was related code.  I agree that it is cleaner to merge these separately.  In practice I'm not sure how important that is here given that this is a new feature branch anyway?  Not sure which ptychography changes you mean.

I mentioned in the PR comment that the documentation was incomplete....so, yup, you are right, it does need to be finished :)  And yes......this requires updating the i/o functionality as well before it will be ready to merge into dev!

**Pull request merged**